### DISCUSSION TITLE
How to overlay multiple scenes with the same camera for display

### DISCUSSION THREAD

**Discussion by WREQI** [pmndrs/postprocessing#630] [Q&A]
How to overlay multiple scenes with the same camera for display

**Comment by WREQI** [comment]
![image](https://github.com/pmndrs/postprocessing/assets/35467304/b7798127-cc0e-4891-99b0-e903c155bba2)
I did this without displaying the output to the screen for the content in the first scene

**Comment by WREQI** [comment]
![image](https://github.com/pmndrs/postprocessing/assets/35467304/e770b378-7fae-4a7e-9184-3f7cd45712d9)
After rendering in Composer, a new scene appears with aliasing, and anti aliasing smaa has been added

**Comment by vanruesc** [comment]
> How to overlay multiple scenes with the same camera

That depends on how you define overlay.

You could try rendering one scene and then saving that image to another render target using a `CopyPass`. Then you could render the other scene and finally use a `TextureEffect` to blend the previously rendered image with the current scene.

**Comment by WREQI** [comment]
I don't quite understand. This did not successfully display two scenes
![image](https://github.com/pmndrs/postprocessing/assets/35467304/e10a6e3a-1cad-4704-ace1-5409fbe419a1)

**Reply by vanruesc** [reply]
The texture for the `TextureEffect` should be `copyPass.texture` and you'll probably want to use `BlendFunction.ALPHA`. The `TextureEffect` needs to be added to an `EffectPass`:

```ts
const effectPass = new EffectPass(camera, textureEffect);

composer.addPass(renderPass1); // Render scene that should be blended with another scene
composer.addPass(copyPass); // Save the result
composer.addPass(renderPass2); // Render the other scene
composer.addPass(effectPass); // Blend the saved result with the current image
```

Setting `renderToScreen` is not necessary because the last pass will automatically render to screen unless `composer.autoRenderToScreen` is set to `false`.

**Comment by WREQI** [comment]
![image](https://github.com/pmndrs/postprocessing/assets/35467304/86a781e7-c109-4977-a385-aff24c653168)
![image](https://github.com/pmndrs/postprocessing/assets/35467304/58c0cd37-2b23-4992-be75-fc175810eb8a)
![image](https://github.com/pmndrs/postprocessing/assets/35467304/523fbeea-4a97-401a-8fea-0c5e53cd4094)
Thank you, according to your method, it was successfully displayed, but the effect is very different from the actual expected effect. 1. After adding post-processing, the object color becomes distorted. I tried setting the BlendFunction function to add, but the normal is not correct. 2. All objects in the scene Helper have perspective penetrating through the objects in the scene

**Reply by WREQI** [reply]
The last image is the original effect I want (without post-processing)

**Comment by WREQI** [comment]
@vanruesc

**Comment by vanruesc** [comment]
I'm sorry, but it's not clear to me what result you're trying to achieve.

If you're simply trying to render two scenes separately, one after another, then you can chain two `RenderPass` instances like you've done in the first screenshot. In that case, you must disable the internal clear pass of the second `RenderPass`:

```js
const renderPass1 = new RenderPass(...);
const renderPass2 = new RenderPass(...);
renderPass2.clearPass.enabled = false;

composer.addPass(renderPass1);
composer.addPass(renderPass2);
composer.addPass(effectPass);
```

**Comment by WREQI** [comment]
![image](https://github.com/pmndrs/postprocessing/assets/35467304/9cfb4c3b-1e25-4458-82cd-31d14a3824ba)
![image](https://github.com/pmndrs/postprocessing/assets/35467304/bf0cc546-70c6-430c-880e-823c246f3ee9)
Color distortion, The color in Figure 1 is incorrect Figure 2 is the real color

**Comment by WREQI** [comment]
The edges will have obvious jagged edges, and the rotating arrows will disappear
![image](https://github.com/pmndrs/postprocessing/assets/35467304/a91f62cb-8ce3-4502-a834-80d1201a7d44)
this demo
https://stackblitz.com/edit/three-ez-postprocessing-external-pzgzg6?file=src%2Fmain.ts

**Comment by WREQI** [comment]
@vanruesc

**Comment by vanruesc** [comment]
If you render a blue arrow on a blue background, you won't be able to see it. The green arrow is slightly transparent, so rendering that on a blue background will yield green-blue results.

The following example uses MSAA to reduce aliasing artifacts. Please note that raising the outline edge strength above 1 will interfere with antialiasing because it expects colors to be in low dynamic range [0.0, 1.0]:

https://stackblitz.com/edit/three-ez-postprocessing-external-2zdftd?file=src%2Fmain.ts

This demo first renders the main scene. Then it applies the outline effect to that scene and finally the transform gizmo is rendered on top of that. The result is then copied to screen.

**Comment by WREQI** [comment]
I added a new box and plane to Scene2, but it encountered rendering errors @vanruesc 
![image](https://github.com/pmndrs/postprocessing/assets/35467304/d62834ab-209e-45af-9ed9-454117811445)
https://stackblitz.com/edit/three-ez-postprocessing-external-kk7sr4?file=src%2Fmain.ts
